pydantic/_internal/_dataclasses.py

```diff
@@ -135,16 +146,13 @@ def __init__(__dataclass_self__: PydanticDataclass, *args: Any, **kwargs: Any) -
             'Support for `__post_init_post_parse__` has been dropped, the method will not be called', DeprecationWarning
         )
 
-    if types_namespace is None:
-        types_namespace = _typing_extra.merge_cls_and_parent_ns(cls)
```

Comment:
I'm not too happy with having both the `ns_resolver` and `parent_namespace` argument in this `complete_dataclass` function. `parent_namespace` is only relevant when doing the initial dataclass build (similarly to Pydantic models -- though currently it isn't provided; see the TODO @ `pydantic/dataclasses.py:L273`).

`ns_resolver` is only relevant when calling `rebuild_dataclass` (it will include a fallback and override namespace). But unlike Pydantic models, a dataclass rebuild will also trigger a `collect_dataclass_fields`. iirc we talked about this inconsistency on Slack, I'm postponing it as it can be decoupled from this PR.

So to sum things up:
- initial build: only the parent ns is at play
- dc rebuild: only the ns_resolver is at play.

Reply:
Resolving, this seems to now be fixed.